File: security-intel-hub/security_intel_hub/collectors/x.py
from __future__ import annotations

from datetime import datetime

from security_intel_hub.http import request_json
from security_intel_hub.models import IntelItem


class XRecentSearchCollector:
    name = "x"

    def __init__(
        self,
        *,
        bearer_token: str,
        query: str,
        max_results: int = 25,
        api_base: str = "https://api.x.com/2",
    ) -> None:
        self.bearer_token = bearer_token
        self.query = query
        self.max_results = max(10, min(max_results, 100))
        self.api_base = api_base.rstrip("/")

    def collect(self) -> list[IntelItem]:
        if not self.bearer_token:
            return []

        response = request_json(
            "GET",
            f"{self.api_base}/tweets/search/recent",
            headers={"Authorization": f"Bearer {self.bearer_token}"},
            params={
                "query": self.query,
                "max_results": self.max_results,
                "tweet.fields": "created_at,author_id,public_metrics,entities",
                "expansions": "author_id",
                "user.fields": "username,name",
            },
        )

        users = {
            user["id"]: user
            for user in response.get("includes", {}).get("users", [])
            if "id" in user
        }
        items: list[IntelItem] = []
        for tweet in response.get("data", []):
            author = users.get(tweet.get("author_id", ""), {})
            username = author.get("username", "")
            url = f"https://x.com/{username}/status/{tweet['id']}" if username else ""
            text = tweet.get("text", "").strip()
            items.append(
                IntelItem(
                    source=self.name,
                    source_id=tweet["id"],
                    title=_headline(text),
                    body=text,
                    url=url,
                    author=username or tweet.get("author_id", ""),
                    published_at=_parse_datetime(tweet.get("created_at")),
                    raw=tweet,
                )
            )
        return items


def _headline(text: str, length: int = 96) -> str:
    text = " ".join(text.split())
    return text if len(text) <= length else f"{text[: length - 1]}..."


def _parse_datetime(value: str | None) -> datetime | None:
    if not value:
        return None
    return datetime.fromisoformat(value.replace("Z", "+00:00"))

File: backend/src/sources/x.ts
import type { FetchResult, ThreatIndicator } from '../types.js';
import { errorMessage, fetchWithTimeout } from '../util.js';
import { clampLimit, headline, parseIsoTime, scoreSocialText } from './social.js';

const DEFAULT_QUERY =
  '(CVE OR exploit OR 0day OR zero-day OR ransomware OR malware OR phishing OR APT OR IOC OR KEV) lang:en -is:retweet';

interface XTweet {
  id: string;
  text?: string;
  author_id?: string;
  created_at?: string;
  public_metrics?: {
    retweet_count?: number;
    reply_count?: number;
    like_count?: number;
    quote_count?: number;
  };
}

interface XUser {
  id: string;
  username?: string;
  name?: string;
}

interface XResponse {
  data?: XTweet[];
  includes?: { users?: XUser[] };
}

function metricBonus(metrics: XTweet['public_metrics']): number {
  if (!metrics) return 0;
  const reposts = (metrics.retweet_count ?? 0) + (metrics.quote_count ?? 0);
  const likes = metrics.like_count ?? 0;
  const replies = metrics.reply_count ?? 0;
  return (
    Math.min(15, Math.floor(Math.log1p(reposts) * 2)) +
    Math.min(10, Math.floor(Math.log1p(likes))) +
    Math.min(5, Math.floor(Math.log1p(replies)))
  );
}

export async function fetchXRecentSearch(): Promise<FetchResult<ThreatIndicator>> {
  const fetchedAt = Date.now();
  const bearerToken = process.env.X_BEARER_TOKEN?.trim();
  if (!bearerToken) return { items: [], fetchedAt, error: null };

  try {
    const apiBase = (process.env.X_API_BASE?.trim() || 'https://api.x.com/2').replace(/\/+$/, '');
    const query = process.env.X_QUERY?.trim() || DEFAULT_QUERY;
    const maxResults = clampLimit(process.env.X_MAX_RESULTS, 25, 10, 100);
    const params = new URLSearchParams({
      query,
      max_results: String(maxResults),
      'tweet.fields': 'created_at,author_id,public_metrics,entities',
      expansions: 'author_id',
      'user.fields': 'username,name',
    });

    const res = await fetchWithTimeout(
      `${apiBase}/tweets/search/recent?${params.toString()}`,
      { headers: { Authorization: `Bearer ${bearerToken}` } },
      30_000,
    );
    if (!res.ok) throw new Error(`HTTP ${res.status}`);

    const data = (await res.json()) as XResponse;
    const users = new Map((data.includes?.users ?? []).map((user) => [user.id, user]));
    const items: ThreatIndicator[] = [];
    for (const tweet of data.data ?? []) {
      const text = tweet.text?.trim();
      if (!text) continue;
      const user = tweet.author_id ? users.get(tweet.author_id) : undefined;
      const username = user?.username;
      const url = username
        ? `https://x.com/${username}/status/${tweet.id}`
        : `https://x.com/i/web/status/${tweet.id}`;
      const scored = scoreSocialText(text, metricBonus(tweet.public_metrics));
      items.push({
        id: `x:${tweet.id}`,
        source: 'x',
        type: 'social_intel',
        indicator: url,
        indicatorType: 'url',
        severity: scored.severity,
        title: headline(text),
        description: text,
        tags: ['x', ...(username ? [`@${username}`] : []), ...scored.tags],
        reference: url,
        firstSeen: parseIsoTime(tweet.created_at),
      });
    }

    return { items, fetchedAt, error: null };
  } catch (err) {
    return { items: [], fetchedAt, error: errorMessage(err) };
  }
}
